lib/autocomplete-manager.coffee

```diff
@@ -272,18 +186,166 @@ class AutocompleteManager
+
+      if range.intersectsWith(selectionRange)
+        prefixOffset = selectionRange.start.column - range.start.column
+        prefix = match[0][0...prefixOffset] if range.start.isLessThan(selectionRange.start)
+
+    return prefix
+
+  # Private: Checks whether the current file is blacklisted.
+  #
+  # Returns {Boolean} that defines whether the current file is blacklisted
+  isCurrentFileBlackListed: ->
+    blacklist = (atom.config.get('autocomplete-plus.fileBlacklist') or '')
+      .split(',')
+      .map((s) -> s.trim())
+
+    fileName = path.basename(@editor.getBuffer().getPath())
+    for blacklistGlob in blacklist
+      return true if minimatch(fileName, blacklistGlob)
+
+    return false
+
+  # Private: Gets called when the content has been modified
+  contentsModified: =>
+    delay = parseInt(atom.config.get('autocomplete-plus.autoActivationDelay'))
+    clearTimeout(@delayTimeout) if @delayTimeout
+    @delayTimeout = setTimeout(@runAutocompletion, delay)
+
+  # Private: Gets called when the cursor has moved. Cancels the autocompletion if
+  # the text has not been changed.
+  #
+  # data - An {Object} containing information on why the cursor has been moved
+  cursorMoved: (data) =>
+    @hideSuggestionList() unless data.textChanged
+
+  editorHasFocus: =>
+    return false unless @editorView?
+    editorView = @editorView
+    editorView = editorView[0] if editorView.jquery
+    return editorView.hasFocus()
+
+  # Private: Gets called when the user saves the document. Cancels the
+  # autocompletion.
+  bufferSaved: =>
+    return unless @editorHasFocus()
+    @hideSuggestionList()
+
+  # Private: Cancels the autocompletion if the user entered more than one
+  # character with a single keystroke. (= pasting)
   #
-  # currentBuffer - The current {TextBuffer}
-  setCurrentBuffer: (@currentBuffer) ->
-    @compositeDisposable.add @currentBuffer.onDidSave(@editorSaved)
-    @compositeDisposable.add @currentBuffer.onDidChange(@editorChanged)
+  # e - The change {Event}
+  bufferChanged: (e) =>
+    return if @suggestionList.compositionInProgress
+    return unless @editorHasFocus()
+    if atom.config.get('autocomplete-plus.enableAutoActivation') and (e.newText.trim().length is 1 or e.oldText.trim().length is 1)
+      @contentsModified()
+    else
+      @hideSuggestionList()
+
+  #  |||              |||
+  #  vvv PROVIDER API vvv
+
+  registerProviderForGrammars: (provider, grammars) =>
+    return unless provider?
+    return unless grammars? and _.size(grammars) > 0
+    grammars = _.filter(grammars, (grammar) -> grammar?.scopeName?)
+    scopes = _.pluck(grammars, 'scopeName')
+    return @registerProviderForScopes(provider, scopes)
+
+  registerProviderForScopes: (provider, scopes) =>
+    return unless provider?
+    return unless scopes? and _.size(scopes) > 0
+    for scope in scopes
+      existing = _.findWhere(_.keys(@scopes), scope)
+      if existing? and @scopes[scope]?
```

Comment:
existing? implies the key exists. @scopes[scope]? implies the value is defined. It's possible - although highly unlikely - it's undefined.

This will likely change when we switch to selectors instead of descriptors.